src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,510 @@
+// error if the input is too short to contain a credential.
+func getCredentialLen(ser []byte) (int, error) {
+	if len(ser) < 6 {
+		// TODO: change this error
+		return 0, errors.New("invalid delegated credential")
+	}
+	// First 4 bytes is the validity time.
+	ser = ser[4:]
+
+	// The next 2 bytes are the length of the serialized public key.
+	serPublicKeyLen := int(binary.BigEndian.Uint16(ser))
+	ser = ser[2:]
+
+	if len(ser) < serPublicKeyLen {
+		return 0, errors.New("public key of credential is too short")
+	}
+
+	return 6 + serPublicKeyLen, nil
+}
+
+// getHash maps the SignatureScheme to its corresponding hash function.
+// TODO: replace this with typeAndHashFromSignatureScheme
+func getHash(scheme SignatureScheme) crypto.Hash {
+	switch scheme {
+	case ECDSAWithP256AndSHA256:
+		return crypto.SHA256
+	case ECDSAWithP384AndSHA384:
+		return crypto.SHA384
+	case ECDSAWithP521AndSHA512:
+		return crypto.SHA512
+	default:
+		return 0 // Unknown hash function
+	}
+}
+
+// getCurve maps the SignatureScheme to its corresponding elliptic.Curve.
+// TODO: replace this with typeAndHashFromSignatureScheme
+func getCurve(scheme SignatureScheme) elliptic.Curve {
+	switch scheme {
+	case ECDSAWithP256AndSHA256:
+		return elliptic.P256()
+	case ECDSAWithP384AndSHA384:
+		return elliptic.P384()
+	case ECDSAWithP521AndSHA512:
+		return elliptic.P521()
+	default:
+		return nil
+	}
+}
+
+// prepareDelegation returns a hash of the message that the delegator is to
+// sign. The inputs are the credential ('cred'), the DER-encoded delegator
+// certificate ('delegatorCert'), the signature scheme of the delegator
+// ('delegatorScheme'), and the protocol version ('vers') in which the credential
+// is to be used.
+// TODO: does this need the hash?
+func prepareDelegation(hash crypto.Hash, cred *credential, dCert []byte, dScheme SignatureScheme, peer string) ([]byte, error) {
+	h := hash.New()
+
+	h.Write(bytes.Repeat([]byte{0x20}, 64))
+
+	if peer == "server" {
+		h.Write([]byte("TLS, server delegated credentials"))
+	} else if peer == "client" {
+		h.Write([]byte("TLS, client delegated credentials"))
+	}
+
+	h.Write([]byte{0x00})
+	h.Write(dCert)
+
+	serCred, err := cred.marshal()
+	if err != nil {
+		return nil, err
+	}
+	h.Write(serCred)
+
+	var serScheme [2]byte
+	binary.BigEndian.PutUint16(serScheme[:], uint16(dScheme))
+	h.Write(serScheme[:])
+
+	return h.Sum(nil), nil
+}
+
+// NewDelegatedCredential creates a new delegated credential using 'cert' for
+// delegation. It generates a public/private key pair for the provided signature
+// algorithm ('scheme'), validity interval (defined by 'cert.Leaf.notBefore' and
+// 'validTime'), and TLS version ('vers'), and signs it using 'cert.PrivateKey'.
+// TODO: add the other signature schemes
+func NewDelegatedCredential(cert *Certificate, scheme SignatureScheme, validTime time.Duration, peer string) (*DelegatedCredential, crypto.PrivateKey, error) {
```

Comment:
It is unclear from the RFC. There are two algorithms to be used:

1. The algorithm used to verify the DC signature, which is the same as the end-entity certificate, as the pub key of it is used for signing.
2. The `expected_cert_verify_algorithm` which is for the delegated credential and should be the same as `CertificateVerify.algorithm` sent by the server, corresponding to the credentials Public Key. 

`scheme SignatureScheme` here corresponds to 2. which should be an input for creating the delegated credential.

Reply:
If the scheme must be equal to the one in the certificate, that means that delegated credentials would not allow to experiment with alternative algorithms for authentication (ie. KEMTLS)

Reply:
The extension in `draft-subcerts` specifies a list of supported signature algorithms for the DC. This, to me, implies that the DC (signature) algorithm can be different from the certificate algorithms.

Reply:
The delegated credential works this way:

1. It has a signature which is generated using the private key of the end-entity certificate
2. It has an `expected_cert_verify_algorithm` that is equal to the DC key pair, and it is the same as the CertificateVerify.algorithm sent by the server

No sure if I follow where the list is. The extension sent in the ClientHello has a list of schemes.

Reply:
The code that iterates over the certificate's Private Key is for signing the DC; but not for the list of signature algorithms that the DC states. 

My answer to the 'does the scheme passed as parameter must be equal to the one used in the certificate?' is that is not clear, in the sense that it can be, but doesn't have to be. That is why it is passed as an argument to the function. Now, this will probably need to be changed to something else, as the KEM will not be of the form of `SignatureScheme`.

Reply:
I hope these comments make it more understandable of how DC works.. let me know if you have any other questions. I'm closing this one.